torch/_inductor/codecache.py

```diff
@@ -761,25 +761,55 @@ def get_include_and_linking_paths(
-            # There is either Intel OpenMP(for x86) or LLVM OpenMP (for both x86 and arm64)
-            libs = ["omp"]
-            if os.getenv("OMP_PREFIX") is not None:
-                # Support OpenMP on MacOS
-                ipaths.append(os.path.join(os.getenv("OMP_PREFIX"), "include"))
-                lpaths.append(os.path.join(os.getenv("OMP_PREFIX"), "lib"))
-
-            if os.getenv("CONDA_PREFIX") is not None:
-                # On MacOS OpenMP is not available via the system install
-                # But on conda can be provided using https://anaconda.org/anaconda/llvm-openmp
-                conda_lib_path = os.path.join(os.getenv("CONDA_PREFIX"), "lib")
-                ipaths.append(os.path.join(os.getenv("CONDA_PREFIX"), "include"))
-                lpaths.append(conda_lib_path)
-                # Prefer Intel OpenMP on x86 machine
-                if os.uname().machine == "x86_64" and os.path.exists(
-                    os.path.join(conda_lib_path, "libiomp5.dylib")
-                ):
-                    libs = ["iomp5"]
+            cxx = cpp_compiler()
+            version_string = subprocess.check_output([cxx, "--version"]).decode('utf8')
+            # only Apple builtin compilers (Apple Clang++) require openmp
+            omp_available = 'Apple' not in version_string.splitlines()[0]
+
+            libs = [] if omp_available else ["omp"]
+
+            # check the `OMP_PREFIX` environment first
+            if not omp_available and os.getenv("OMP_PREFIX") is not None:
+                header_path = os.path.join(os.getenv("OMP_PREFIX"), "include", "omp.h")
+                omp_available = os.path.exists(header_path)
+                if omp_available:
+                    ipaths.append(os.path.join(os.getenv("OMP_PREFIX"), "include"))
+                    lpaths.append(os.path.join(os.getenv("OMP_PREFIX"), "lib"))
+
+            # prefer to use openmp from `conda install llvm-openmp`
+            if not omp_available and os.getenv("CONDA_PREFIX") is not None:
+                command = "conda list llvm-openmp --json"
+                output = subprocess.check_output(command.split()).decode('utf8')
+                omp_available = len(json.loads(output)) > 0
+                if omp_available:
+                    conda_lib_path = os.path.join(os.getenv("CONDA_PREFIX"), "lib")
+                    ipaths.append(os.path.join(os.getenv("CONDA_PREFIX"), "include"))
+                    lpaths.append(conda_lib_path)
+                    # Prefer Intel OpenMP on x86 machine
+                    if os.uname().machine == "x86_64" and os.path.exists(
+                        os.path.join(conda_lib_path, "libiomp5.dylib")
+                    ):
+                        libs = ["iomp5"]
+
+            # next, try to use openmp from `brew install libomp`
+            if not omp_available:
+                try:
+                    # check if `brew` is installed
+                    subprocess.check_output(["which", "brew"])
```

Comment:
We should cache results of subprocess calls.

Reply:
All cached in [78ac245](https://github.com/pytorch/pytorch/pull/107111/commits/78ac245f0ee34d0647972db4f212c8bff4f8e459).